### ISSUE TITLE
Claude-generated: backward() after an infeasible or unbounded solve silently returns finite gradients

### ISSUE THREAD

**Issue by sbarratt** [moreau-project/moreau#59]
Claude-generated. Found by an external review of 0.4.1 and re-verified on main @ e794100 (CPU).

## Problem

Calling backward after a solve whose status is `PrimalInfeasible`, `DualInfeasible` (or any non-solved status) silently returns finite gradients, in torch and JAX, for both solvers. There is no solution to differentiate, so these numbers are meaningless, and the two solvers disagree on the same problem. Training loops continue on a constant loss (the portfolio doc example does exactly this, see #68).

```python
import numpy as np, torch, moreau
from moreau.torch import Solver
# x >= 1 and x <= 0 (infeasible);  min 1/2 x^2
for kind in ("ipm", "active_set"):
    s = Solver(1, 2, torch.tensor([0, 1]), torch.tensor([0]), torch.tensor([0, 1, 2]), torch.tensor([0, 0]),
               moreau.Cones(num_nonneg_cones=2), settings=moreau.Settings(device="cpu", solver=kind))
    f64 = lambda a, g=False: torch.tensor(a, dtype=torch.float64, requires_grad=g)
    q, b = f64([0.0], True), f64([-1.0, 0.0], True)
    s.solve(f64([1.0]), f64([-1.0, 1.0]), q, b).x.sum().backward()
    print(kind, s.info.status, b.grad.numpy())
```

```
ipm        PrimalInfeasible  [9.5e-09 1.0]
active_set PrimalInfeasible  [-0.5 0.5]
```

## Scope

- [ ] Raise (or at minimum warn and return NaN) when backward is requested for a problem whose status isn't `Solved`/`AlmostSolved`, in torch, JAX, and `Solver.backward`. For batches, do it per problem.
- [ ] Document the behaviour in the differentiation guide.
